yarn/src/main/scala/org/apache/spark/deploy/yarn/AMDelegationTokenRenewer.scala

```diff
@@ -0,0 +1,203 @@
+                "hour! If this happens too often tasks will fail.", e)
+              delegationTokenRenewer.schedule(this, 1, TimeUnit.HOURS)
+              return
+          }
+          scheduleRenewal(this)
+        }
+      }
+    // Schedule update of credentials. This handles the case of updating the tokens right now
+    // as well, since the renenwal interval will be 0, and the thread will get scheduled
+    // immediately.
+    scheduleRenewal(driverTokenRenewerRunnable)
+  }
+
+  // Keeps only files that are newer than 30 days, and deletes everything else. At least 5 files
+  // are kept for safety
+  private def cleanupOldFiles(): Unit = {
+    import scala.concurrent.duration._
+    try {
+      val remoteFs = FileSystem.get(hadoopConf)
+      val credentialsPath = new Path(sparkConf.get("spark.yarn.credentials.file"))
+      val thresholdTime = System.currentTimeMillis() - (30 days).toMillis
+      hadoopUtil.listFilesSorted(
+        remoteFs, credentialsPath.getParent, credentialsPath.getName, ".tmp")
+        .dropRight(5)
+        .takeWhile(_.getModificationTime < thresholdTime)
+        .foreach(x => remoteFs.delete(x.getPath, true))
+    } catch {
+      // Such errors are not fatal, so don't throw. Make sure they are logged though
+      case e: Exception =>
+        logWarning("Error while attempting to cleanup old tokens. If you are seeing many such " +
+          "warnings there may be an issue with your HDFS cluster.")
+    }
+  }
+
+  private def writeNewTokensToHDFS(principal: String, keytab: String): Unit = {
+    if (!loggedInViaKeytab) {
+      // Keytab is copied by YARN to the working directory of the AM, so full path is
+      // not needed.
+      logInfo(s"Attempting to login to KDC using principal: $principal")
+      UserGroupInformation.loginUserFromKeytab(principal, keytab)
+      logInfo("Successfully logged into KDC.")
+      loggedInViaKeytab = true
+    }
+    val nns = YarnSparkHadoopUtil.get.getNameNodesToAccess(sparkConf)
+    hadoopUtil.obtainTokensForNamenodes(
+      nns, hadoopConf, UserGroupInformation.getCurrentUser.getCredentials)
+    val remoteFs = FileSystem.get(hadoopConf)
+    // If lastCredentialsFileSuffix is 0, then the AM is either started or restarted. If the AM
+    // was restarted, then the lastCredentialsFileSuffix might be > 0, so find the newest file
+    // and update the lastCredentialsFileSuffix.
+    if (lastCredentialsFileSuffix == 0) {
+      val credentialsPath = new Path(sparkConf.get("spark.yarn.credentials.file"))
+      hadoopUtil.listFilesSorted(
+        remoteFs, credentialsPath.getParent, credentialsPath.getName, ".tmp")
+        .lastOption.foreach { status =>
+        lastCredentialsFileSuffix = getSuffixForCredentialsPath(status)
+      }
+    }
+    val nextSuffix = lastCredentialsFileSuffix + 1
+    val tokenPathStr =
+      sparkConf.get("spark.yarn.credentials.file") + "-" + nextSuffix
```

Comment:
It would be nice to make some utility functions for doing construction and destruction of the names with "-". It would also be nice to have a constant for the .tmp we use everywhere.